pulp/apis/sas_api.py

```diff
@@ -0,0 +1,657 @@
+                    results="TEXT",
+                )
+
+            self.delete_tmp_files(tmpMps, tmpMst)
+
+            # Store SAS output
+            if self.msg:
+                print(res["LOG"])
+            self._macro = dict(
+                (key.strip(), value.strip())
+                for key, value in (
+                    pair.split("=") for pair in sas.symget("_OR" + proc + "_").split()
+                )
+            )
+
+            primal_out = sas.sd2df(f"primalout{postfix}")
+            dual_out = sas.sd2df(f"dualout{postfix}")
+
+            if self._macro.get("STATUS", "ERROR") != "OK":
+                raise PulpSolverError(
+                    "PuLP: Error ({err_name}) \
+                        while trying to solve the instance: {name}".format(
+                        err_name=self._macro.get("STATUS", "ERROR"), name=lp.name
+                    )
+                )
+            status = self._read_solution(lp, primal_out, dual_out)
+
+            return status
+
+
+class SASCAS(SASsolver):
+    name = "SASCAS"
+
+    try:
+        global swat
+        import swat
+
+    except ImportError:
+
+        def available(self):
+            """True if SASCAS is available."""
+            return False
+
+        def sasAvailable(self):
+            return False
+
+        def actualSolve(self, lp, callback=None):
+            """Solves a well-formulated lp problem."""
+            raise PulpSolverError("SASCAS : Not Available")
+
+    else:
+
+        def __init__(
+            self,
+            mip=True,
+            msg=True,
+            keepFiles=False,
+            warmStart=False,
+            timeLimit=None,
+            cas=None,
+            **solverParams,
+        ):
+            """
+            :param bool mip: if False, assume LP even if integer variables
+            :param bool msg: if False, no log is shown
+            :param bool keepFiles: if False, mps and mst files will not be saved
+            :param bool warmStart: if False, no warmstart or initial primal solution provided
+            :param cas: CAS object. See swat.CAS
+            :param solverParams: SAS proc OPTMILP or OPTLP parameters
+            """
+            SASsolver.__init__(
+                self,
+                mip=mip,
+                msg=msg,
+                keepFiles=keepFiles,
+                warmStart=warmStart,
+                timeLimit=timeLimit,
+                **solverParams,
+            )
+            self.cas = cas
+
+        def available(self):
+            return True
+
+        def sasAvailable(self):
+            if self.cas == None:
+                return False
+            try:
+                with redirect_stdout(SASLogWriter(self.msg)) as self._log_writer:
+                    # Load the optimization action set
+                    self.cas.loadactionset("optimization")
+                    return True
+            except:
+                return False
+
+        def toDict(self):
```

Comment:
(1) you have this function duplicated.
(2) you should not just copy the inherit method, you should call the `LpSolver_CMD.toDict` method and then add the functionality that you need.
(3) What is cas object and how do you expect to serialize it?